Adding containsAnyOf assertion

marko-bekhta opened the pull request:
#### Check List:
* Fixes #954
* Unit tests : YES 
* Javadoc with a code example (API only) : NO

Hey, that's my first PR here, so if the approach is correct I can add JavaDocs and anything else if missing. Thanks

joel-costigliola commented on src/main/java/org/assertj/core/internal/Iterables.java:
```diff
@@ -917,6 +919,18 @@ public void assertContainsExactly(AssertionInfo info, Iterable<?> actual, Object
                            shouldContainExactly(actual, values, diff.missing, diff.unexpected, comparisonStrategy));
   }
 
+  public void containsAnyOf(AssertionInfo info, Iterable<?> actual, Object[] values) {
```
rename to `assertContainsAnyOf` for consistency with other `Iterables` method.

joel-costigliola commented on src/main/java/org/assertj/core/internal/Iterables.java:
```diff
@@ -917,6 +919,18 @@ public void assertContainsExactly(AssertionInfo info, Iterable<?> actual, Object
 
+  public void containsAnyOf(AssertionInfo info, Iterable<?> actual, Object[] values) {
+    checkIsNotNull(values);
+    assertNotNull(info, actual);
```
I believe these checks can be replaced by `commonCheckThatIterableAssertionSucceeds` that would also handle the empty case.

joel-costigliola commented on src/main/java/org/assertj/core/internal/Iterables.java:
```diff
@@ -917,6 +919,18 @@ public void assertContainsExactly(AssertionInfo info, Iterable<?> actual, Object
+    assertNotNull(info, actual);
+    Set<Object> valuesToSearchFor = newTreeSet(values);
+    for (Object element : actual) {
+      if (valuesToSearchFor.contains(element)) {
```
if possible  in your IDE, remove curly braces and put `return;` in the same line

joel-costigliola commented on src/main/java/org/assertj/core/internal/Iterables.java:
```diff
@@ -917,6 +919,18 @@ public void assertContainsExactly(AssertionInfo info, Iterable<?> actual, Object
+        return;
+      }
+    }
+    throw failures.failure(info, ShouldContainAnyOf.shouldContainAnyOf(actual, values));
```
static import `shouldContainAnyOf` for readability

joel-costigliola commented:
@marko-bekhta thanks for the PR ! 
I have reviewed it and I have a few additional comments that did not fit in the PR review: 

- can you add unit tests for `assertContainsAnyOf` in `Iterables_assertContainsAnyOf_Test.java` as in [Iterables_assertContains_Test.java](https://github.com/joel-costigliola/assertj-core/blob/52aa454c7ac2150741b0e98289dbefe1e0b9653d/src/test/java/org/assertj/core/internal/iterables/Iterables_assertContains_Test.java), please try to cover all the corner cases.
- if you have the time, can you add `containsAnyElementsOf` ? it is the same assertion but it takes an `Iterable` in place of a vararg.

marko-bekhta commented:
@joel-costigliola thanks for the review. I'll try to address all the things that you've mentioned later today.

marko-bekhta commented:
Changed the things that were mentioned. In addition, changed parameter names to be consistent with others,  and added JavaDocs, used `Iterables#iterableContains` instead of regular `Set#contains`.

Also found one missing param name in the documentation - so added it as well.

joel-costigliola commented on src/test/java/org/assertj/core/internal/iterables/Iterables_assertContainsAnyOf_Test.java:
```diff
@@ -0,0 +1,155 @@
+  }
+
+  @Test
+  public void should_fail_if_actual_does_not_contain_any_of_values() {
```
`should_fail_if_actual_does_not_contain_any_of_the_given_values`

joel-costigliola commented on src/main/java/org/assertj/core/api/AbstractIterableAssert.java:
```diff
@@ -677,6 +678,64 @@ public SELF usingDefaultElementComparator() {
+   */
+  public SELF containsAnyElementsOf(Iterable<ELEMENT> iterable) {
+    iterables.assertContainsAnyElementsOf(info, actual, iterable);
+    return myself;
```
it is simpler to call `containsAnyOf` like this: `return containsAnyOf(toArray(iterable));` then the only test you need for this guys is that it delegates correctly to `iterables.assertContainsAnyOf`.
You can also get rid of `iterables.assertContainsAnyElementsOf` and its corresponding tests.

marko-bekhta replied:
addressed this and updated the commit.

joel-costigliola commented:
Integrated thank you @marko-bekhta  !
